A new Promise-based testing API

Turbo87 opened the pull request:
I'd like to propose a substantial API change for these helpers that should resolve #38, simplify the whole process of writing assertions and also greatly simplify our own code in this project.

Some examples:

```js
  it('route foo', function() {
    var args = ['route', 'foo'];

    return emberNew()
      .then(() => emberGenerate(args))
      .then(() => expect(file('app/routes/foo.js'))
        .to.contain('import Ember from \'ember\';')
        .to.contain('export default Ember.Route.extend({\n});'))

      .then(() => expect(file('app/templates/foo.hbs'))
        .to.contain('{{outlet}}'))

      .then(() => expect(file('tests/unit/routes/foo-test.js'))
        .to.contain('import { moduleFor, test } from \'ember-qunit\';')
        .to.contain('moduleFor(\'route:foo\''))

      .then(() => expect(file('app/router.js'))
        .to.contain('this.route(\'foo\')'))

      .then(() => emberDestroy(args))
      .then(() => expect(file('app/routes/foo.js')).to.not.exist)
      .then(() => expect(file('app/templates/foo.hbs')).to.not.exist)
      .then(() => expect(file('tests/unit/routes/foo-test.js')).to.not.exist)
      .then(() => expect(file('app/router.js'))
        .to.not.contain('this.route(\'foo\')'));
  });
```

```js
  it('adapter throws when --base-class is same as name', function() {
    var args = ['adapter', 'foo', '--base-class=foo'];

    return emberNew()
      .then(() => expect(emberGenerate(args))
        .to.be.rejectedWith(SilentError, /Adapters cannot extend from themself/));
  });
```

with async/await:

```js

  it('route foo', async function() {
    var args = ['route', 'foo'];

    await emberNew();
    await emberGenerate(args);

    expect(file('app/routes/foo.js'))
      .to.contain('import Ember from \'ember\';')
      .to.contain('export default Ember.Route.extend({\n});');

    expect(file('app/templates/foo.hbs'))
      .to.contain('{{outlet}}');

    expect(file('tests/unit/routes/foo-test.js'))
      .to.contain('import { moduleFor, test } from \'ember-qunit\';')
      .to.contain('moduleFor(\'route:foo\'');

    expect(file('app/router.js'))
      .to.contain('this.route(\'foo\')');

    await emberDestroy(args);

    expect(file('app/routes/foo.js')).to.not.exist;
    expect(file('app/templates/foo.hbs')).to.not.exist;
    expect(file('tests/unit/routes/foo-test.js')).to.not.exist;
    expect(file('app/router.js'))
      .to.not.contain('this.route(\'foo\')');
  });
```

I've experimented with this API already in the legacy blueprints project and you can find my converted code at https://github.com/Turbo87/ember-cli-legacy-blueprints/commits/bp-v2.

We can probably support both APIs for a certain amount of time to make the transition easier for everyone that's using the current API.

Note that this is work-in-progress and not everything is polished yet, but the general concept seems to work very well.

/cc @trabus @stefanpenner @rwjblue @kellyselden @nathanhammond

nathanhammond commented:
One complaint with syntax, this feels awkward:
```
.to.contain('import Ember from \'ember\';').and
.to.contain('export default Ember.Route.extend({\n});'))
```

How does this API work with `await`/`async`? I want that to feel good in some future state. (Yes, this is a `node` test that isn't transpiled, but we'll get that at some point and we should plan for it.)

trabus commented:
@nathanhammond I agree, and had suggested maybe implementing a few helpers that could bulk assert (taking an array of contents or files). Regardless, the assertion style should be a preference of the person writing the tests, we just need to provide the right places for them to be. However, I agree that if we're going to go with babel, we might as well go ahead and try and align with the testing unification RFC and somehow implement async/await.

@Turbo87 I'm still of the opinion that there is value in the `genrateAndDestroy` helper, as it provides a very simple interface that doesn't require the user to worry about writing any assertions themselves. I feel like the flow you present is better for asserting specific things outside the normal `generateAndDestroy`, but should really only be used for those special cases.

Turbo87 commented:
@nathanhammond the `.and` chain in those assertions is optional and can be left out as well. 

I've experimented with async/await and it works really well and simplifies the syntax even further. I've added an async/await example in the PR description above. (see also https://github.com/Turbo87/ember-cli-legacy-blueprints/blob/async-await/node-tests/blueprints/route-test.js#L23-L69)

One drawback of using Babel for the async/await transpilation however is that the resulting stacktrace is entirely useless and doesn't point you to the failing assertion at all.

nathanhammond commented:
@Turbo87 I like the `async`/`await` syntax. As long as we're planning for its future default existence in V8 I don't think we need to adopt it via Babel. Pushing toward modern `node` is a much easier proposal than trying to move an entire client-side ecosystem, so I don't mind waiting for it.

*As long as it shows up soon.

stefanpenner commented:
* looks good
* seems like all the chai helpers could remain sync
* I'm nervous about using babel in node, unless the source-map situation for stack traces in tests works nicely. If it does (i haven't tried in ages)

stefanpenner commented on lib/chai-files.js:
```diff
@@ -0,0 +1,108 @@
+var fs = require('fs');
```
i would love access to this chai helper from other places, could I convince you this may want to live as its own module?

Turbo87 replied:
it should definitely live in it's own module. I just used this place to for a first draft that can be extracted out later. the error messages can also still be improved.

Turbo87 replied:
extracted into https://github.com/Turbo87/chai-files

Turbo87 commented:
> seems like all the chai helpers could remain sync

they are sync. I just felt that the code looks a bit more structured and aligned when they're wrapped in `.then()` calls too

> I'm nervous about using babel in node, unless the source-map situation for stack traces in tests works nicely. If it does (i haven't tried in ages)

unfortunately it doesn't work that well. it seems to work okay as long as we're only using arrow functions, but transpiling async/await completely breaks the stacktraces as mentioned above. I would propose using the arrow function syntax for now and once node gains native async/await support we could transition to that instead.

Turbo87 commented:
my goal with this PR was to get some feedback if this would go in the right direction or if I'm wasting my time on it. since the feedback is largely positive I can now start refining the code. 

/cc @stefanpenner

stefanpenner commented:
@Turbo87 I like the ergonomic improvements here  :+1:

Turbo87 commented:
I've updated the branch with a `emberGenerateDestroy()` convenience function and rewrote the README file according to the new API.

Examples from the new README:

```js
describe('Acceptance: ember generate and destroy my-blueprint', function() {
  // create and destroy temporary working directories
  setupTestHooks(this);

  it('my-blueprint foo', function() {
    var args = ['my-blueprint', 'foo'];

    // create a new Ember.js app in the working directory
    return emberNew()
      
      // then generate and destroy the `my-blueprint` blueprint called `foo`
      .then(() => emberGenerateDestroy(args, (file) => {

        // and run some assertions in between
        expect(file('path/to/file.js'))
          .to.contain('file contents to match')
          .to.contain('more file contents\n');
      }));
      
     // magically done for you: assert that the generated files are destroyed again 
  });
});
```

or more explicitly:

```js
describe('Acceptance: ember generate and destroy my-blueprint', function() {
  // create and destroy temporary working directories
  setupTestHooks(this);

  it('my-blueprint foo', function() {
    var args = ['my-blueprint', 'foo'];

    // create a new Ember.js app in the working directory
    return emberNew()
      
      // then generate the `my-blueprint` blueprint called `foo`
      .then(() => emberGenerate(args))

      // then assert that the files were generated correctly
      .then(() => expect(file('path/to/file.js'))
        .to.contain('file contents to match')
        .to.contain('more file contents\n'))
      
      // then destroy the `my-blueprint` blueprint called `foo`
      .then(() => emberDestroy(args))

      // then assert that the files were destroyed correctly
      .then(() => expect(file('path/to/file.js')).to.not.exist);
  });
});
```

mmun commented:
This looks great to me.

stefanpenner commented:
LGTM

trabus commented:
This is great work! Onward to 1.0.0! :rocket:
